optuna/_callbacks.py

```diff
@@ -0,0 +1,59 @@
+from typing import Tuple
+
+import optuna
+from optuna._experimental import experimental
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+@experimental("2.8.0")
+class MaxTrialsCallback:
+    """Set a maximum number of trials before ending the study.
+
+    While the :obj:`n_trials` argument of :obj:`optuna.optimize` sets the number of trials that
+    will be run, you may want to continue running until you have a certain number of successfullly
+    completed trials or stop the study when you have a certain number of trials that fail.
+    This :obj:`MaxTrialsCallback` class allows you to set a maximum number of trials for a
+    particular :class:`~optuna.trial.TrialState` before stopping the study.
+
+    Example:
+
+        .. testcode::
+
+            import optuna
+            from optuna.study import MaxTrialsCallback
+            from optuna.trial import TrialState
+
+
+            def objective(trial):
+                x = trial.suggest_float("x", -1, 1)
+                return x ** 2
+
+
+            study = optuna.create_study()
+            study.optimize(
+                objective,
+                callbacks=[MaxTrialsCallback(10, states=(TrialState.COMPLETE,))],
+            )
+
+    Args:
+        n_trials:
+            The max number of trials. Must be set to an integer.
+        states:
+            Tuple of the :class:`~optuna.trial.TrialState` to be counted
+            towards the max trials limit. Default value is :obj:`(TrialState.COMPLETE,)`.
+    """
+
+    def __init__(
+        self, n_trials: int, states: Tuple[TrialState, ...] = (TrialState.COMPLETE,)
+    ) -> None:
+        self._n_trials = n_trials
+        self._states = states
+
+    def __call__(
+        self, study: "optuna.study.Study", trial: FrozenTrial
+    ) -> None:  # NOQA # type: ignore
```

Comment:
Again, we shouldn't need the exceptions here. https://github.com/optuna/optuna/pull/2612#discussion_r631557813

Reply:
Right. 😢

Reply:
Thanks ❤️